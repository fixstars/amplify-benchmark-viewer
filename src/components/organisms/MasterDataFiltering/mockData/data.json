[
  { "name": "Tsp" },
  { "name": "Qap" },
  { "name": "MaxCut" },
  { "name": "Sudoku" },
  { "name": "Cvrp" }
]
