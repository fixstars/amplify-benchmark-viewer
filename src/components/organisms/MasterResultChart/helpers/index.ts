import type { MasterData } from '@types'
import type { Data, Layout } from 'plotly.js'

const getLayout = (bestKnown: number): Partial<Layout> => {
  return {
    updatemenus: [
      {
        buttons: [
          {
            args: [{}, { xaxis: { title: 'sampling_time[ms]', type: 'log' } }],
            label: 'Log Scale',
            method: 'update',
          },
          {
            args: [
              {},
              {
                xaxis: {
                  title: 'sampling_time[ms]',
                  type: 'linear',
                },
              },
            ],
            label: 'Linear Scale',
            method: 'update',
          },
        ],
        xanchor: 'left',
        y: 1.2,
        yanchor: 'top',
      },
    ],
    xaxis: {
      title: 'sampling_time[ms]',
      type: 'log',
      autorange: true,
    },
    yaxis: {
      title: 'target_energy',
      autorange: true,
    },
    showlegend: true,
    shapes: [
      {
        type: 'line',
        xref: 'paper',
        x0: 0,
        y0: bestKnown,
        x1: 1,
        y1: bestKnown,
        line: {
          dash: 'dash',
        },
      },
    ],
  }
}

interface PlotDataParameter {
  readonly data: ReadonlyArray<MasterData>
  readonly clientChartType: {
    readonly [key: string]: string
  }
}

const getPlotData = ({
  data,
  clientChartType,
}: PlotDataParameter): {
  readonly plotData: { readonly [key: string]: Data[] }
  readonly plotOriginData: { readonly [key: string]: MasterData }
  readonly problemBestKnown: { readonly [key: string]: number }
} => {
  const plotGroupIDList = []
  for (const item of data) {
    const groupID = item.group_id
    if (groupID === '') continue

    plotGroupIDList.push(groupID)
  }

  const problemBestKnown: { [key: string]: number } = {}
  const plotData: { [key: string]: Data[] | undefined } = {}
  const plotOriginData: { [key: string]: MasterData } = {}
  for (const item of data) {
    const groupID = item['group_id']
    if (plotGroupIDList.includes(groupID)) {
      const type = clientChartType[item['client_name']]
      const instance = item['problem_instance']
      problemBestKnown[instance] = item['problem_best_known']

      if ((item['plot_data'][type] as string | null) == null) continue

      const index = (plotData[instance]?.length ?? 0) + 1
      const name = `${index}_${item['client_name']}(${item['client_version']})`
      item['plot_data'][type]['name'] = name
      plotOriginData[`${name}_${instance}`] = item

      if (instance in plotData) {
        plotData[instance]?.push(item['plot_data'][type])
      } else {
        plotData[instance] = [item['plot_data'][type]]
      }
    }
  }

  return {
    plotData: plotData as { readonly [key: string]: Data[] },
    plotOriginData,
    problemBestKnown,
  }
}

export { getLayout, getPlotData }
