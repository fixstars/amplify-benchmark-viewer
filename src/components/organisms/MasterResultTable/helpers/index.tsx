import { Tooltip } from '@mui/material'
import type { GridColumnGroupingModel, GridColumns } from '@mui/x-data-grid'
import type { MasterData } from '@types'

const renderCell = (row: MasterData, field: string) => {
  return (
    <Tooltip
      title={
        <>
          {row.client_parameters != null && (
            <>
              <span>Client parameters</span>
              <br />
              {Object.entries(row.client_parameters).map(
                ([key, value], index) => (
                  <div key={`${key}-${index}`}>
                    <span>- {key}: </span>
                    <span>{`${value}`}</span>
                  </div>
                ),
              )}
            </>
          )}
          {row.problem_parameters != null && (
            <>
              {row.client_parameters != null && <br />}
              <span>Problem parameters</span>
              <br />
              {Object.entries(row.problem_parameters).map(
                ([key, value], index) => (
                  <div key={`${key}-${index}`}>
                    <span>- {key}: </span>
                    <span>{`${value}`}</span>
                  </div>
                ),
              )}
            </>
          )}
        </>
      }
    >
      <span>{row[field as never]}</span>
    </Tooltip>
  )
}

const columns: GridColumns<MasterData> = [
  {
    field: 'problem_class',
    headerName: 'Problem class',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'problem_instance',
    headerName: 'Problem instance',
    sortable: false,
    minWidth: 140,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'client_name',
    headerName: 'Client name',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'client_version',
    headerName: 'Client version',
    sortable: false,
    minWidth: 200,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'setting_time',
    headerName: 'Setting time',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'num_samples',
    headerName: 'Num samples',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'num_feasibles',
    headerName: 'Num feasibles',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'num_reach_best',
    headerName: 'Num reach best',
    sortable: false,
    minWidth: 140,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'min',
    headerName: 'Min',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: '25%',
    headerName: '25%',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: '50%',
    headerName: '50%',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: '75%',
    headerName: '75%',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
  {
    field: 'max',
    headerName: 'Max',
    sortable: false,
    minWidth: 120,
    renderCell: (params) => renderCell(params.row, params.field),
  },
]

const columnGroups: GridColumnGroupingModel = [
  {
    groupId: 'Basic information',
    children: [
      { field: 'problem_class' },
      { field: 'problem_instance' },
      { field: 'client_name' },
      { field: 'client_version' },
      { field: 'setting_time' },
      { field: 'num_samples' },
      { field: 'num_feasibles' },
      { field: 'num_reach_best' },
    ],
  },
  {
    groupId: 'Target energy',
    children: [
      { field: 'min' },
      { field: '25%' },
      { field: '50%' },
      { field: '75%' },
      { field: 'max' },
    ],
  },
]

export { columns, columnGroups }
