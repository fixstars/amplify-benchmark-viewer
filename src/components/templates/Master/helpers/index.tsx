import type { FilteringData, MasterData } from '@types'

interface FilteringOptions {
  readonly problemClassData: readonly FilteringData[]
  readonly clientNameData: readonly FilteringData[]
  readonly clientVersionData: readonly FilteringData[]
}

const createFilteringData = (
  data: ReadonlyArray<MasterData>,
): FilteringOptions => {
  const problemClassData: FilteringData[] = []
  const clientNameData: FilteringData[] = []
  const clientVersionData: FilteringData[] = []

  data.forEach(
    ({
      problem_class: problemClass,
      client_name: clientName,
      client_version: clientVersion,
    }) => {
      let index = problemClassData.findIndex(
        (problem) => problem.name === problemClass,
      )
      if (index < 0) {
        problemClassData.push({
          name: problemClass,
        })
      }

      index = clientNameData.findIndex((client) => client.name === clientName)
      if (index < 0) {
        clientNameData.push({
          name: clientName,
        })
      }

      index = clientVersionData.findIndex(
        (client) => client.name === clientVersion,
      )
      if (index < 0) {
        clientVersionData.push({
          name: clientVersion,
        })
      }
    },
  )

  return {
    problemClassData,
    clientNameData,
    clientVersionData,
  }
}

interface Filters {
  readonly data: readonly MasterData[]
  readonly problemClassFilter: ReadonlyArray<string>
  readonly clientNameFilter: ReadonlyArray<string>
  readonly clientVersionFilter: ReadonlyArray<string>
  readonly problemInstanceFilter?: string
}

const getFilteringData = ({
  data,
  problemClassFilter,
  clientNameFilter,
  clientVersionFilter,
  problemInstanceFilter,
}: Filters): MasterData[] => {
  return data
    .filter(({ problem_class: problemClass }) => {
      return problemClassFilter.length > 0
        ? problemClassFilter.includes(problemClass)
        : true
    })
    .filter(({ client_name: clientName }) => {
      return clientNameFilter.length > 0
        ? clientNameFilter.includes(clientName)
        : true
    })
    .filter(({ client_version: clientVersion }) => {
      return clientVersionFilter.length > 0
        ? clientVersionFilter.includes(clientVersion)
        : true
    })
    .filter(({ problem_instance: problemInstance }) => {
      return problemInstanceFilter != null
        ? problemInstance === problemInstanceFilter
        : true
    })
}

interface ChartOptions {
  // eslint-disable-next-line functional/prefer-readonly-type
  [key: string]: Set<string> | null
}

const getChartOptionsByClients = (
  data: readonly MasterData[],
): ChartOptions => {
  const options: ChartOptions = {}

  for (const element of data) {
    if (options[element['client_name']] == null) {
      options[element['client_name']] = new Set<string>()
    }

    for (const [key] of Object.entries(element['plot_data'])) {
      options[element['client_name']]?.add(key)
    }
  }

  return options
}

const getProblemInstances = (
  data: readonly MasterData[],
): ReadonlyArray<string> => {
  const problemInstances = new Set<string>()

  for (const element of data) {
    problemInstances.add(element.problem_instance)
  }

  return Array.from(problemInstances)
}

export {
  createFilteringData,
  getFilteringData,
  getChartOptionsByClients,
  getProblemInstances,
}
