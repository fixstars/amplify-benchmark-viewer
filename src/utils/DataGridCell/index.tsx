/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { LinkButton } from 'components/atoms/LinkButton'

const clientColors = [
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
]

const instanceColors = [
  '#ff66ff',
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#718de6',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
  '#9a916d',
  '#406a0f',
  '#004b31',
  '#18729f',
  '#9997c3',
  '#9cf0ec',
  '#ce843c',
  '#2d326e',
  '#76ed60',
  '#d7f7e8',
  '#bb15d7',
  '#87a7d8',
  '#94053a',
  '#e5c8a6',
  '#e2f2c0',
]

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    const cache = JSON.parse(localStorage.getItem('clientColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('clientColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[client]

    if (backgroundColor === undefined) {
      const length = clientColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = clientColors[cacheIndex]
      cache[client] = backgroundColor
      localStorage.setItem('clientColorCache', JSON.stringify(cache))
      localStorage.setItem('clientColorIndexCache', JSON.stringify(cacheIndex))
      return backgroundColor
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (instance: string): string => {
    const cache = JSON.parse(localStorage.getItem('instanceColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('instanceColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[instance]

    if (backgroundColor === undefined) {
      const length = instanceColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = instanceColors[cacheIndex]
      cache[instance] = backgroundColor
      localStorage.setItem('instanceColorCache', JSON.stringify(cache))
      localStorage.setItem(
        'instanceColorIndexCache',
        JSON.stringify(cacheIndex),
      )
      return backgroundColor
    }

    return backgroundColor
  }
  const pairs = Object.entries(problemInstances)
  pairs.sort(function (p1, p2) {
    const p1Key = p1[0]
    const p2Key = p2[0]
    if (p1Key < p2Key) {
      return -1
    }
    if (p1Key > p2Key) {
      return 1
    }
    return 0
  })
  problemInstances = Object.fromEntries(pairs)
  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {problemInstances[className]?.map((instance) => (
          <div
            key={`${key}_${className}_${instance}`}
            style={{ marginRight: 10, marginBottom: 10 }}
          >
            <LinkButton
              label={instance}
              link={`/classes/${className}/instances/${instance}`}
              backgroundColor={getBackgroundColor(instance)}
            />
          </div>
        ))}
      </div>
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
