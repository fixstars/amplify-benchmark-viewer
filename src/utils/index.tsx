export * from './DataGridCell'
export * from './statistics'
