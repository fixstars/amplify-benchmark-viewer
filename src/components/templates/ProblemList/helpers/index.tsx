import type { GridColumns } from '@mui/x-data-grid'
import { Link } from 'react-router-dom'
import { renderClientsCell } from 'utils/DataGridCell'

export interface ProblemClients {
  readonly class: string
  readonly instance: string
  readonly clients: ReadonlyArray<string>
}

const renderInstanceCell = (row: ProblemClients) => {
  return (
    <Link to={`/classes/${row.class}/instances/${row.instance}`}>
      {row.instance}
    </Link>
  )
}

const columns: GridColumns<ProblemClients> = [
  {
    field: 'class',
    headerName: 'Class',
    sortable: false,
  },
  {
    field: 'instance',
    headerName: 'Instance',
    sortable: false,
    minWidth: 180,
    renderCell: (params) => renderInstanceCell(params.row),
  },
  {
    field: 'clients',
    headerName: 'Clients',
    sortable: false,
    flex: 1,
    renderCell: ({ row }) =>
      renderClientsCell(`${row.class}_${row.instance}`, row.clients),
  },
]

export { columns }
