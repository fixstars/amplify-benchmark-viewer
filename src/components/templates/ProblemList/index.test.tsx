/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { mockLayout } from 'utils/test'
import mockData from 'utils/test/mocks/data/mockProblemListData.json'

import { ProblemList } from '.'
import { columns } from './helpers'

describe('<ProblemList />', () => {
  it('Rendered well', async () => {
    const { container } = render(
      <BrowserRouter>
        <ProblemList data={mockData} />
      </BrowserRouter>,
    )

    const layout = mockLayout.mock.calls[0][0]
    expect(layout.children.length).toBe(2)

    const sectionTitleContainer = layout.children[0]
    expect(sectionTitleContainer.type.name).toBe('Box')
    const sectionTitle = sectionTitleContainer.props.children
    expect(sectionTitle.type.name).toBe('SectionTitle')
    expect(sectionTitle.props.title).toBe('Problem List')

    const contentsContainer = layout.children[1]
    expect(contentsContainer.type.name).toBe('Box')
    expect(contentsContainer.props.sx).toEqual({ display: 'flex', flex: 1 })
    const dataGrid = contentsContainer.props.children
    expect(dataGrid.type.name).toBe('DataGrid')
    expect(dataGrid.props.sx).toEqual({
      '&.MuiDataGrid-root .MuiDataGrid-cell:focus-within': {
        outline: 'none !important',
      },
    })
    expect(dataGrid.props.rows).toEqual(mockData)
    expect(dataGrid.props.columns).toEqual(columns)
    expect(dataGrid.props.getRowId.name).toBe('getRowId')
    expect(dataGrid.props.pagination).toBe(true)
    expect(dataGrid.props.disableColumnMenu).toBe(true)
    expect(dataGrid.props.initialState).toEqual({
      pagination: {
        paginationModel: {
          pageSize: 100,
        },
      },
    })

    expect(container).toMatchSnapshot()
  })
})
