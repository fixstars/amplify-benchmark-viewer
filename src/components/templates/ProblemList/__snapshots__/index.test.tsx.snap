// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProblemList /> Rendered well 1`] = `
<div>
  <div
    class="MuiBox-root css-thz6x0"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1hc7nu0-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1r339ym-MuiToolbar-root"
      >
        <div
          class="css-yp9swi"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-19sbe58-MuiTypography-root"
            >
              Annealing Benchmark
            </h6>
          </a>
        </div>
        <nav
          class="css-ov1ktg"
        >
          <a
            class="css-94l1l1"
            href="/"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-4hrd4f-MuiTypography-root"
            >
              Problems
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/clients"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Clients
            </p>
          </a>
          <a
            class="css-94l1l1"
            href="/labels"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-muv6ew-MuiTypography-root"
            >
              Label
            </p>
          </a>
        </nav>
      </div>
    </header>
    <div
      class="MuiBox-root css-1cf7osc"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-15h1z56-MuiGrid-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h2 css-1asg1ln-MuiTypography-root"
          >
            Problem List
          </h2>
        </div>
      </div>
      <div
        class="MuiBox-root css-13pmxen"
      >
        <div
          aria-colcount="3"
          aria-multiselectable="false"
          aria-rowcount="8"
          class="MuiDataGrid-root MuiDataGrid-root--densityStandard css-15i6hkc-MuiDataGrid-root"
          role="grid"
        >
          <div>
            <div />
          </div>
          <div
            class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
          >
            <div
              class="MuiDataGrid-columnHeaders css-f3jnds-MuiDataGrid-columnHeaders"
              style="min-height: 56px; max-height: 56px; line-height: 56px;"
            >
              <div
                class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
                role="rowgroup"
                style="transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-rowindex="1"
                  class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
                  role="row"
                >
                  <div
                    aria-colindex="1"
                    aria-label="Class"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader"
                    data-field="class"
                    role="columnheader"
                    style="height: 56px; width: 100px; min-width: 100px; max-width: 100px;"
                    tabindex="0"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Class
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="2"
                    aria-label="Instance"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader"
                    data-field="instance"
                    role="columnheader"
                    style="height: 56px; width: 180px; min-width: 180px; max-width: 180px;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Instance
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    aria-colindex="3"
                    aria-label="Clients"
                    aria-sort="none"
                    class="MuiDataGrid-columnHeader"
                    data-field="clients"
                    role="columnheader"
                    style="height: 56px; width: 0px; min-width: 0; max-width: 0;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderDraggableContainer"
                      draggable="false"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainer"
                      >
                        <div
                          class="MuiDataGrid-columnHeaderTitleContainerContent"
                        >
                          <div
                            aria-label=""
                            class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                            data-mui-internal-clone-element="true"
                          >
                            Clients
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                      style="min-height: 56px; opacity: 1;"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                        data-testid="SeparatorIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M11 19V5h2v14z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              style="overflow: visible; height: 0px; width: 0px;"
            >
              <div
                class="MuiDataGrid-virtualScroller css-1w5m2wr-MuiDataGrid-virtualScroller"
                style="width: 0px; height: auto; margin-top: 56px; overflow-x: hidden;"
              >
                <div
                  class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
                  style="width: auto; height: 364px; min-height: auto;"
                >
                  <div
                    class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                    style="transform: translate3d(0px, 0px, 0px);"
                  >
                    <div
                      aria-rowindex="2"
                      aria-selected="false"
                      class="MuiDataGrid-row"
                      data-id="TSP_s_instance"
                      data-rowindex="0"
                      role="row"
                      style="max-height: 52px; min-height: 52px;"
                    >
                      <div
                        aria-colindex="1"
                        aria-colspan="1"
                        class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="0"
                        data-field="class"
                        role="cell"
                        style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <div
                          class="MuiDataGrid-cellContent"
                        >
                          TSP
                        </div>
                      </div>
                      <div
                        aria-colindex="2"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="1"
                        data-field="instance"
                        role="cell"
                        style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <a
                          href="/classes/TSP/instances/s_instance"
                        >
                          s_instance
                        </a>
                      </div>
                      <div
                        aria-colindex="3"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="2"
                        data-field="clients"
                        role="cell"
                        style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-1o7gueb"
                            href="/clients/a_client"
                          >
                            a_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-gqb4bp"
                            href="/clients/b_client"
                          >
                            b_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-196ffja"
                            href="/clients/c_client"
                          >
                            c_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-14frf9o"
                            href="/clients/d_client"
                          >
                            d_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-tudjox"
                            href="/clients/e_client"
                          >
                            e_client
                          </a>
                        </div>
                      </div>
                    </div>
                    <div
                      aria-rowindex="3"
                      aria-selected="false"
                      class="MuiDataGrid-row"
                      data-id="TSP_t_instance"
                      data-rowindex="1"
                      role="row"
                      style="max-height: 52px; min-height: 52px;"
                    >
                      <div
                        aria-colindex="1"
                        aria-colspan="1"
                        class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="0"
                        data-field="class"
                        role="cell"
                        style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <div
                          class="MuiDataGrid-cellContent"
                        >
                          TSP
                        </div>
                      </div>
                      <div
                        aria-colindex="2"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="1"
                        data-field="instance"
                        role="cell"
                        style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <a
                          href="/classes/TSP/instances/t_instance"
                        >
                          t_instance
                        </a>
                      </div>
                      <div
                        aria-colindex="3"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="2"
                        data-field="clients"
                        role="cell"
                        style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-1o7gueb"
                            href="/clients/a_client"
                          >
                            a_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-gqb4bp"
                            href="/clients/b_client"
                          >
                            b_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-14frf9o"
                            href="/clients/d_client"
                          >
                            d_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-tudjox"
                            href="/clients/e_client"
                          >
                            e_client
                          </a>
                        </div>
                      </div>
                    </div>
                    <div
                      aria-rowindex="4"
                      aria-selected="false"
                      class="MuiDataGrid-row"
                      data-id="QAP_u_instance"
                      data-rowindex="2"
                      role="row"
                      style="max-height: 52px; min-height: 52px;"
                    >
                      <div
                        aria-colindex="1"
                        aria-colspan="1"
                        class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="0"
                        data-field="class"
                        role="cell"
                        style="min-width: 100px; max-width: 100px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <div
                          class="MuiDataGrid-cellContent"
                        >
                          QAP
                        </div>
                      </div>
                      <div
                        aria-colindex="2"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="1"
                        data-field="instance"
                        role="cell"
                        style="min-width: 180px; max-width: 180px; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <a
                          href="/classes/QAP/instances/u_instance"
                        >
                          u_instance
                        </a>
                      </div>
                      <div
                        aria-colindex="3"
                        aria-colspan="1"
                        class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                        data-colindex="2"
                        data-field="clients"
                        role="cell"
                        style="min-width: 0; max-width: 0; min-height: 52px; max-height: 52px;"
                        tabindex="-1"
                      >
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-1o7gueb"
                            href="/clients/a_client"
                          >
                            a_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-196ffja"
                            href="/clients/c_client"
                          >
                            c_client
                          </a>
                        </div>
                        <div
                          style="margin-right: 10px;"
                        >
                          <a
                            class="css-14frf9o"
                            href="/clients/d_client"
                          >
                            d_client
                          </a>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="Mui-resizeTriggers"
            >
              <div
                class="expand-trigger"
              >
                <div
                  style="width: 1px; height: 1px;"
                />
              </div>
              <div
                class="contract-trigger"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
